File: app/src/main/java/com/rikvanvelzen/codingtest/common/DIold/presentation/ViewModelModule.java
/*
 * Created by Rik van Velzen, Norakomi Software Development.
 * Copyright (c) 2020. All rights reserved
 * Last modified 1/30/20 1:05 PM
 */

package com.rikvanvelzen.codingtest.common.DIold.presentation;


import androidx.lifecycle.ViewModel;

import com.rikvanvelzen.codingtest.common.DIold.viewmodel.ViewModelFactory;
import com.rikvanvelzen.codingtest.ui.screens.currencies.CurrencyViewModel;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.util.Map;

import javax.inject.Provider;

import dagger.MapKey;
import dagger.Module;
import dagger.Provides;
import dagger.multibindings.IntoMap;

@Module
public class ViewModelModule {

    @Target(ElementType.METHOD)
    @Retention(RetentionPolicy.RUNTIME)
    @MapKey
    @interface ViewModelKey {
        Class<? extends ViewModel> value();
    }

    @Provides
    ViewModelFactory viewModelFactory(Map<Class<? extends ViewModel>, Provider<ViewModel>> providerMap) {
        return new ViewModelFactory(providerMap);
    }

    @Provides
    @IntoMap
    @ViewModelKey(CurrencyViewModel.class)
    ViewModel currencyViewModel() {
        return new CurrencyViewModel();
    }

}

File: app/src/main/java/com/rikvanvelzen/codingtest/common/dependencyinjection/modules/ViewModelModule.kt
package com.rikvanvelzen.codingtest.common.dependencyinjection.modules

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.rikvanvelzen.codingtest.ui.screens.currencies.CurrencyViewModel
import com.rikvanvelzen.tbocodingchallenge.common.dependencyinjection.ViewModelFactory
import com.rikvanvelzen.tbocodingchallenge.common.dependencyinjection.ViewModelKey
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
abstract class ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(CurrencyViewModel::class)
    abstract fun bindCurrencyViewModel(viewModel: CurrencyViewModel): ViewModel

    @Binds
    abstract fun bindViewModelFactory(viewModelFactory: ViewModelFactory): ViewModelProvider.Factory
}